### PR TITLE
Create Create F2_21.c

### PR DESCRIPTION
**Pull request by italo-07**
The application you will be writing for is intended to run on a small, inexpensive embedded device that does not have a floating point unit. This means that the processor can only do integer arithmetic. However, the device does have a display and needs to display the result of floating point arithmetic to the user. Because this is a small inexpensive processor there is no support for strings or any mathematical library functions. You must write all the code yourselves performing all the floating point math with integers only and no strings! Character arrays, or C strings, can't tell you their size but they do end with a '\0' character.

The requirements for the code are to write one of the functions presented below. All the functions return a boolean value indicating the success or failure of the operation. Improper data is the most likely cause for a function to fail. You may add your own additional helper functions (I had at least a half dozen) but you must include them for the review. Do not change the interface to these functions.

These two functions take as parameters two sets of mantissa and characteristic and a char array to hold the result of the arithmetic operations. The result of the add or subtract should be converted into char's and placed in the result array. The array must end with a '\0'. The 'len' parameter tells how many characters can be placed on the result array. For these functions to return true you must at a minimum store the characteristic of the result. If the result is a non-integer place as many of the digits as will fit in the result after a decimal point.

bool add(int c1, int n1, int d1, int c2, int n2, int d2, char result[], int len);
bool subtract(int c1, int n1, int d1, int c2, int n2, int d2, char result[], int len); 
...
...
char answer[100];
int c1, n1, d1;
int c2, n2, d2;
 
c1 = 1;
n1 = 1;
d1 = 2;
 
c2 = 2;
n2 = 2;
d2 = 3; 
 
//if the C string could hold at least the characteristic if(add(c1, n1, d1, c2, n2, d2, answer, 100))
{
    //display string with answer 4.166666666...
}
else
{
    //display error message
}
'''

### CODE REVIEW

**Comment by arjoyita** [comment]
I’m reviewing this PR now.

**Review by arjoyita — COMMENTED** [review]
**Summary:** 
Clear problem framing and constraints (integer-only math, no strings/libs, '\0'-terminated output, fixed buffer len). The function interfaces are appropriate for an embedded target, and the success/failure contract is explicit. Nice job calling out that at minimum the characteristic must fit for a true return.

**What’s working well:**
1.Interfaces are minimal and testable:
bool add(int c1,int n1,int d1,int c2,int n2,int d2,char result[],int len) and subtract(...).
2.Constraints are realistic for MCUs without FPU/stdlib.
3.Output contract is specific: decimal point + as many digits as fit.

**Key issues / required clarifications**
1.Valid input checks: define behavior for d1==0 || d2==0 (must return false). Clarify if negative n1/n2 are permitted or whether sign must be in c1/c2 only.
2.Normalization & reduction: spell out required normalization (e.g., ensure 0 ≤ n < d post-normalize) and always reduce the final fraction via gcd before formatting.
3.Overflow handling: with common denominators like d1*d2, intermediate values can overflow int. Either (a) use careful cross-addition with early reduction, (b) widen to long long for intermediates, or (c) document that inputs are small enough to avoid overflow.
4.Formatting spec: specify rounding vs truncation. The text implies truncate after the decimal to whatever fits; confirm that no rounding is required.
5.len contract: define the minimal len that must succeed (e.g., enough to print at least the characteristic and '\0'). Explicitly state that the function returns false if even that can’t fit.
6.Negative values: define the placement of '-' (before the characteristic) and how it interacts with len (if len only fits "-\0", return false).
7.Tests: add a few table-driven examples (incl. edge cases):
a) 1 1/2 + 2 2/3 = 4.1666…
b) -1 3/4 + 0 1/2 = -1.25
c) Very small len (1, 2, 3), zero denominators, large numerators approaching overflow.

**Suggested implementation outline (non-library, integer-only)**
1.Helpers (no stdlib, no strings):
   +static int igcd(int a,int b);
   +static void norm_sign(int *c,int *n,int *d); // move sign to characteristic
   +static int write_int(char *dst,int cap,int v); // returns written chars (or -1)
   +static int write_fraction_decimal(char *dst,int cap,int num,int den); // long-division style, truncate
2.Algorithm:
    +Validate inputs (d1>0 && d2>0, len>0, result!=NULL).
    +Normalize both operands to improper fractions in a widened type (e.g., long long): x = c*d + sign(n)*n.
    +Add/sub via num = x1*d2 ± x2*d1, den = d1*d2; reduce by gcd.
    +Split to mixed form: C = num/den, rem = abs(num%den).
    +Emit sign, C, optional '.', then decimal digits by repeated (rem*=10; digit=rem/den; rem%=den) until buffer fills; always '\0' terminate; return true only if at least C fit.

**Style & embedded concerns**
1. Avoid recursion; prefer small, static helpers.
2. No dynamic memory.
3. Keep worst-case loops bounded by len to avoid unbounded runtime.
4. Document that the function is reentrant (no global state).

File: Create F2_21.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
+Create F2_21.c
+The application you will be writing for is intended to run on a small, inexpensive embedded device that does not have a floating point unit. This means that the processor can only do integer arithmetic. However, the device does have a display and needs to display the result of floating point arithmetic to the user.
```
Comment (arjoyita) [inline Create F2_21.c:2]: This .c file begins with assignment instructions rather than compilable C. Please move the problem description to documentation or a comment header and keep only real C code in the implementation file.

File: Create F2_21.c [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+The requirements for the code are to write one of the functions presented below. All the functions return a boolean value indicating the success or failure of the operation. Improper data is the most likely cause for a function to fail. You may add your own additional helper functions (I had at least a half dozen) but you must include them for the review. Do not change the interface to these functions.
+
+
+These two functions take as parameters two sets of mantissa and characteristic and a char array to hold the result of the arithmetic operations. The result of the add or subtract should be converted into char's and placed in the result array. The array must end with a '\0'. The 'len' parameter tells how many characters can be placed on the result array. For these functions to return true you must at a minimum store the characteristic of the result. If the result is a non-integer place as many of the digits as will fit in the result after a decimal point.
+```
+bool add(int c1, int n1, int d1, int c2, int n2, int d2, char result[], int len);
+bool subtract(int c1, int n1, int d1, int c2, int n2, int d2, char result[], int len); 
+...
```
Comment (arjoyita) [inline Create F2_21.c:12]: These are correct interfaces but they’re currently inside markdown fences/backticks. Please include <stdbool.h> and provide full function bodies in C (not prose). Keep the interface unchanged.

File: Create F2_21.c [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+c1 = 1;
+n1 = 1;
+d1 = 2;
+ 
+c2 = 2;
+n2 = 2;
+d2 = 3; 
+ 
+//if the C string could hold at least the characteristic
+if(add(c1, n1, d1, c2, n2, d2, answer, 100))
```
Comment (arjoyita) [inline Create F2_21.c:29]: Great driver example. Suggest moving this into a small test program or wrapping with #ifdef TEST so the implementation file only contains production code. Also document the buffer policy (no partial writes; result must be '\0'-terminated on success; return false if even sign+characteristic can’t fit).

File: Create F2_21.c [hunk 4]
```diff
@@ -0,0 +1,37 @@
+Create F2_21.c
+The application you will be writing for is intended to run on a small, inexpensive embedded device that does not have a floating point unit. This means that the processor can only do integer arithmetic. However, the device does have a display and needs to display the result of floating point arithmetic to the user. 
+Because this is a small inexpensive processor there is no support for strings or any mathematical library functions. You must write all the code yourselves performing all the floating point math with integers only and no strings! Character arrays, or C strings, can't tell you their size but they do end with a '\0' character.
```
Comment (arjoyita) [inline Create F2_21.c:3]: The spec mentions printing as many digits as fit after the decimal. Please clarify that this means truncation (not rounding). If there’s no space for at least one digit after '.', the function should return false rather than printing alone.
